torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -1368,13 +1439,26 @@ XLAGraphExecutor::SyncTensorsGraphInternal(
   PostOrderData po_data = RunPostOrder(ir_values, &coll);
   coll.hash = torch::lazy::HashCombine(
       coll.hash, torch::lazy::Hash(po_data.parameter_sequence));
+  if (ShouldAliasBasedOnBufferDonor()) {
+    std::vector<size_t> buffer_donor_index =
+        GetBufferDonorIndex(po_data.parameters_data);
+    if (buffer_donor_index.size() > 0) {
+      // Do not include hash on a empty vector.
+      coll.hash = torch::lazy::HashCombine(
+          coll.hash, torch::lazy::Hash(buffer_donor_index));
+    }
+  }
+
   DebugUtil::SaveGraphHash(coll.hash);
   TF_VLOG(4) << "Parameter sequence graph hash "
              << torch::lazy::HashToString(coll.hash);
-  std::shared_ptr<Async> async =
-      TryRunCachedSync(tensors, &coll, &po_data, tensor_data_vec);
-  if (async != nullptr) {
-    return async;
+
+  std::pair<bool, std::shared_ptr<XLAGraphExecutor::Async>> cache_res =
+      TryRunCachedSync(tensors, &coll, &po_data, tensor_data_vec,
+                       warm_up_cache_only);
+  if (cache_res.first) {
+    // we have a cache hit, execution has been scheduled by TryRunCachedSync.
+    return cache_res.second;
```

Comment:
Question: Looks like after adding the `warm_up_cache_only` arg, the `cached_res.second` will be null if `warm_up_cache_only=true`, why do we need this behavior?

Reply:
we do. This is actually kind of funny, from dynamo's persprctive
```
a = a + 1
```
and 
```
a += 1
```
are different so it will generate different dynamo hash and ask XLA to compiles. However from XLA's perspective, the HLO generated for both cases are the same. When `a += 1` compiled by dynamo after `a = a + 1`, it will ask PyTorch/XLA to compile this new graph. Without the change here, we will call `RunCachedGraph` and schedule an execution. The change here was to separate the 
1. check if there is a graph hash hit
2. schedule the graph execution

In the warm_up_cache case, we only want to check if there is a hash match(to avoid recompilation).